ingest-app/src/cmr/ingest/services/granule_bulk_update_service.clj

```diff
@@ -1,79 +1,162 @@
+        updated-metadata (add-opendap-url context concept url)]
+    (if-let [err-messages (:errors updated-metadata)]
+      (errors/throw-service-errors :invalid-data err-messages)
+      (-> concept
+          (assoc :metadata updated-metadata)
+          (update :revision-id inc)
+          (assoc :revision-date (time-keeper/now))
+          (assoc :user-id user-id)))))
+
+(defmethod update-granule-concept :default
+[context concept bulk-update-params user-id]
+)
+
+(defn- update-granule-concept-and-status
+  "Perform update for the granule concept and granule bulk update status."
+  [context task-id concept granule-ur bulk-update-params user-id]
+  (if-let [updated-concept (update-granule-concept context concept bulk-update-params user-id)]
+    (do
+      (ingest-service/save-granule context updated-concept)
+      (data-granule-bulk-update/update-bulk-update-task-granule-status
+       context task-id granule-ur bulk-update-service/updated-status ""))
+    (data-granule-bulk-update/update-bulk-update-task-granule-status
+     context task-id granule-ur bulk-update-service/skipped-status
+     (format (str "Granule with granule-ur [%s] in task-id [%s] is not updated "
+                  "because the metadata format [%s] is not supported.")
+             granule-ur task-id (:format concept)))))
 
 (defn handle-granule-bulk-update-event
-  "Skeleton handler function"
   [context provider-id task-id bulk-update-params user-id]
   (let [{:keys [granule-ur]} bulk-update-params]
-    (data-granule-bulk-update/update-bulk-update-granule-status
-     context
-     task-id
-     granule-ur
-     bulk-update-service/failed-status
-     (format "Granule-ur [%s] is not associated with provider-id [%s]." granule-ur provider-id))))
+    (try
+      (if-let [concept (mdb/find-latest-concept
+                        context {:provider-id provider-id :granule-ur granule-ur} :granule)]
+        (if (:deleted concept)
+          (data-granule-bulk-update/update-bulk-update-task-granule-status
+           context task-id granule-ur bulk-update-service/failed-status
+           (format (str "Granule with granule-ur [%s] on provider [%s] in task-id [%s] "
+                        "is deleted. Can not be updated.")
+                   granule-ur provider-id task-id))
+          ;; granule found and not deleted, update the granule
+          (update-granule-concept-and-status
+           context task-id concept granule-ur bulk-update-params user-id))
+        ;; granule not found
+        (data-granule-bulk-update/update-bulk-update-task-granule-status
+         context task-id granule-ur bulk-update-service/failed-status
+         (format "Granule UR [%s] in task-id [%s] does not exist." granule-ur task-id)))
+      (catch clojure.lang.ExceptionInfo ex-info
+        (error "handle-granule-bulk-update-event caught ExceptionInfo:" ex-info)
+        (if (= :conflict (:type (.getData ex-info)))
+          ;; Concurrent update - re-queue concept update
```

Comment:
I like this. In the future we might want to add a retry count so we don't get stuck in a loop

Reply:
This is a rare race condition that pretty much never happens especially for granules and should be fixed on retry. I am only including this code because collection bulk update code has it.